plugins/analytics-module-ga/src/apis/implementations/AnalyticsApi/GoogleAnalytics.ts

```diff
@@ -0,0 +1,160 @@
+    trackingId: string;
+    testMode: boolean;
+    debug: boolean;
+  }) {
+    this.cdmConfig = cdmConfig;
+
+    // Initialize Google Analytics.
+    ReactGA.initialize(trackingId, { testMode, debug, titleCase: false });
+  }
+
+  /**
+   * Instantiate a fully configured GA Analytics API implementation.
+   */
+  static fromConfig(config: Config) {
+    // Get all necessary configuration.
+    const trackingId = config.getString('app.analytics.ga.trackingId');
+    const debug = !!config.getOptionalBoolean('app.analytics.ga.debug');
+    const testMode = !!config.getOptionalBoolean('app.analytics.ga.testMode');
+    const cdmConfig =
+      config
+        .getOptionalConfigArray('app.analytics.ga.customDimensionsMetrics')
+        ?.map(c => {
+          return {
+            type: c.getString('type') as CDM['type'],
+            index: c.getNumber('index'),
+            source: c.getString('source') as CDM['source'],
+            key: c.getString('key'),
+          };
+        }) || [];
+
+    // Return an implementation instance.
+    return new GoogleAnalytics({
+      trackingId,
+      cdmConfig,
+      testMode,
+      debug,
+    });
+  }
+
+  /**
+   * Primary event capture implementation. Handles core navigate event as a
+   * pageview and the rest as custom events. All custom dimensions/metrics are
+   * applied as they should be (set on pageview, merged object on events).
+   */
+  captureEvent({
+    context,
+    action,
+    subject,
+    value,
+    attributes,
+  }: AnalyticsEvent) {
+    const customMetadata = this.getCustomDimensionMetrics(context, attributes);
+
+    if (action === 'navigate' && context?.extension === 'App') {
+      // Set any/all custom dimensions.
+      if (Object.keys(customMetadata).length) {
+        ReactGA.set(customMetadata);
+      }
+
+      ReactGA.pageview(subject);
+      return;
+    }
+
+    ReactGA.event({
+      category: context.extension || 'App',
+      action,
+      label: subject,
+      value,
+      ...customMetadata,
+    });
+  }
+
+  /**
+   * Returns an object of dimensions/metrics given an Analytics Context and an
+   * Event Context, e.g. { dimension1: "some value", metric8: 42 }
+   */
+  private getCustomDimensionMetrics(
+    context: AnalyticsContextValue,
+    attributes: AnalyticsEventAttributes = {},
+  ) {
+    const dataArray = this.cdmConfig
+      .map(cdm => {
+        const value =
+          cdm.source === 'context' ? context[cdm.key] : attributes[cdm.key];
+
+        // Never pass a non-numeric value on a metric.
+        if (cdm.type === 'metric' && typeof value !== 'number') {
+          return undefined;
+        }
+
+        return value !== undefined
+          ? {
+              [`${cdm.type}${cdm.index}`]: value,
+            }
+          : undefined;
+      })
+      .filter(notEmpty);
+
+    return dataArray.length
+      ? dataArray.reduce((result, cd) => {
+          return Object.assign(result, cd);
+        })
+      : {};
+  }
```

Comment:
I think it can be done in one iteration, something like this:

```ts
 return this.cdmConfig
  .reduce((data, { source, index, type, key }) => {
    const value = source === 'context' ? context[key] : attributes[key];

    // Never pass a non-numeric value on a metric.
    if (type === 'metric' && typeof value !== 'number') {
      return data;
    }
        
    if (value === undefined) {
      return data;
    }
        
    return { ...data, [`${type}${index}`]: value };
  }, {});
```

Reply:
... or with a `for` loop and some explicit `if`s etc :) i'm not a big fan of big complex reductions for readability, pretty sure they turn out to be massively slower as well, in most cases

Reply:
Thanks y'all!  Went with a `forEach()` that just appended keys to an object.  A lot clearer!  And a bit more efficient too, perhaps.